Full screen flashes when resuming from PIP

danae404 opened the pull request:
#### 5a93ec6dc5b9a5575696f4cfac8a5ea778f564e4
<pre>
Full screen flashes when resuming from PIP
<a href="https://bugs.webkit.org/show_bug.cgi?id=286035">https://bugs.webkit.org/show_bug.cgi?id=286035</a>
<a href="https://rdar.apple.com/103974885">rdar://103974885</a>

Reviewed by Jer Noble.

When returning from pip to inline on iPad, the pip placard briefly flashes
before the video content has returned to inline. This is caused by the
placard being in the shadow DOM, instead of part of the UI view hierarchy.

To fix this, the pip placard in modern media controls should not be activated
if the platform is iOS. There is a new pip placard added in VideoPresentationInterfaceIOS.

* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype._shouldControlsBeAvailable):
* Source/WebCore/Modules/modern-media-controls/media/placard-support.js:
(PlacardSupport.prototype._updatePlacard):
(PlacardSupport.prototype._controlsNeedJavascriptPipPlacard):
* Source/WebCore/PAL/pal/ios/UIKitSoftLink.h:
* Source/WebCore/PAL/pal/ios/UIKitSoftLink.mm:
* Source/WebCore/platform/cocoa/WebAVPlayerLayerView.mm:
(WebCore::WebAVPlayerLayerView_transferVideoViewTo):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::ensurePipPlacardIsShowing):
(WebCore::VideoPresentationInterfaceIOS::setupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::preparedToReturnToInline):
(WebCore::VideoPresentationInterfaceIOS::doExitFullscreen):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/289188@main">https://commits.webkit.org/289188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1de21550c6702919e433c4eaa0f9067f9c06130b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4826 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39497 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90244 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36152 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87187 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4916 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66196 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24014 "Found 60 new test failures: animations/3d/animate-to-transform-perpective-to-none.html animations/3d/change-transform-in-end-event.html animations/3d/full-rotation-animation.html animations/3d/replace-filling-transform.html animations/3d/state-at-end-event-transform.html animations/3d/transform-origin-vs-functions.html animations/3d/transform-perspective.html animations/CSSKeyframesRule-name-null.html animations/CSSKeyframesRule-parameters.html animations/added-while-suspended.html ... (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88144 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3746 "Found 2 new test failures: fast/forms/ios/file-upload-panel.html fast/forms/switch/click-animation-twice-fast.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77322 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46463 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3635 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31559 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35225 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74419 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32379 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91639 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12440 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9078 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74709 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12669 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73146 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73823 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18218 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16661 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4454 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13352 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12388 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17869 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12224 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15719 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13970 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -136,13 +140,77 @@ @interface UIViewController ()
+    imageView.get().contentMode = UIViewContentModeScaleAspectFit;
+    imageView.get().translatesAutoresizingMaskIntoConstraints = NO;
+
+    [pipPlacard.get() addSubview:imageView.get() ];
```
`.get()` is unnecessary when calling methods.

pxlcoder commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -136,13 +140,77 @@ @interface UIViewController ()
+        [m_pipPlacard setHidden:NO];
+        return;
+    }
+    auto pipPlacard = adoptNS([PAL::allocUIViewInstance() initWithFrame:[layerHostView() bounds]]);
```
Prefer explicit `RetainPtr` over `auto`, due to smart pointer guidelines. Couple instances throughout the patch.

pxlcoder commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -136,13 +140,77 @@ @interface UIViewController ()
+        return;
+    }
+    auto pipPlacard = adoptNS([PAL::allocUIViewInstance() initWithFrame:[layerHostView() bounds]]);
+    pipPlacard.get().backgroundColor = (UIColor *)[PAL::getUIColorClass() blackColor];
```
We generally use method syntax to avoid having `.get()` all over the code. Couple instances throughout the patch.

In this case you can `[pipPlacard setBackgroundColor:...]`.

Also is the cast really necessary?

danae404 replied:
I think the cast is necessary - I was having compiling trouble without it, and clearUIColor() casts as well (I updated the patch with a grayUIColor() helper function to make the code a bit neater).

pxlcoder commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
```diff
@@ -136,13 +140,77 @@ @interface UIViewController ()
+    pipLabel.get().text = @"This video is playing in picture in picture.";
+    pipLabel.get().textAlignment = NSTextAlignmentCenter;
+    pipLabel.get().textColor = imageColor;
+    pipLabel.get().font = [PAL::getUIFontClass() systemFontOfSize:16];
```
Do you really want a fixed font size, or should this use dynamic type / text styles?

danae404 replied:
I based this decision off of placard.css, where the font size of the text is set to a constant of 13px (I thought 16px looked a bit better)

pxlcoder replied:
You might want to use `UIFontTextStyleCallout` (which is 16 [by default](https://developer.apple.com/design/human-interface-guidelines/typography)), so that the text adapts for users that have [custom accessibility settings](https://support.apple.com/en-us/102453). 

https://developer.apple.com/documentation/uikit/uifont/textstyle/callout?language=objc

danae404 committed:
Full screen flashes when resuming from PIP
https://bugs.webkit.org/show_bug.cgi?id=286035
rdar://103974885

Reviewed by Jer Noble.

When returning from pip to inline on iPad, the pip placard briefly flashes
before the video content has returned to inline. This is caused by the
placard being in the shadow DOM, instead of part of the UI view hierarchy.

To fix this, the pip placard in modern media controls should not be activated
if the platform is iOS. There is a new pip placard added in VideoPresentationInterfaceIOS.

* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype._shouldControlsBeAvailable):
* Source/WebCore/Modules/modern-media-controls/media/placard-support.js:
(PlacardSupport.prototype._updatePlacard):
(PlacardSupport.prototype._controlsNeedJavascriptPipPlacard):
* Source/WebCore/PAL/pal/ios/UIKitSoftLink.h:
* Source/WebCore/PAL/pal/ios/UIKitSoftLink.mm:
* Source/WebCore/platform/cocoa/WebAVPlayerLayerView.mm:
(WebCore::WebAVPlayerLayerView_transferVideoViewTo):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::ensurePipPlacardIsShowing):
(WebCore::VideoPresentationInterfaceIOS::setupFullscreen):
(WebCore::VideoPresentationInterfaceIOS::preparedToReturnToInline):
(WebCore::VideoPresentationInterfaceIOS::doExitFullscreen):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/289188@main
